### ISSUE TITLE
[Design proposal] [In progress] Point in time search 

### ISSUE THREAD

**Issue by bharath-techie** [opensearch-project/OpenSearch#3960]
## Overview
Today, in OpenSearch, if you run the same search query at different points of time, chances are you will get different result as data is constantly changing. However, in real world scenario when analysing data or trying to provide a consistent user experience to your end users you may want the result from a query not to change while the context remains the same and control when changes should appear in the result set. You want to be able to run multiple queries on the same data set and paginate through the data set expecting consistent result. This is not possible using current available options in OpenSearch.

OpenSearch currently supports the following options to achieve pagination, each having a certain limitation:

1. *Scroll API* : Scroll API cannot share point in time context with other queries. Moreover, the scroll API only allows to move forwards(next page) in the search, cases when the client sends the request for a page but fails to get a response, a subsequent retry call skips the page(retried for) and returns the next page in the scroll. 
2. *Search After* : The search_after mechanism doesn't preserve the state of data when the search was issued, so one can paginate using the key (search_after) and fetch subsequent pages while getting more recent results since the search was issued as the pagination progresses.
3. *From To :* This mechanism does not support deep pagination since every page request requires the shard to process all previous results and then filter the requested page which might be taxing deeper the pagination goes

Point in time search can be used to address some of the above shortcomings, 

## What is Point in Time Search
User can create a Point In Time for a set of indices. A Point In Time is a set of segments which are frozen in time taken at the moment of creation. We lock the segments of those indices’ shards and create contexts to access and query only those shards. 
The Create PIT API returns a PIT Id. This PIT ID can be used as part of multiple search queries to get results on that view of data. The indices might ingest more documents or process modification/deletion of documents, but the Point In Time created will be immune to that and provide a view of data which remains unchanged from the time of creation. This allows users to run multiple search queries on a given Point In Time and gives the same result every time.

For feature proposal please refer to - https://github.com/opensearch-project/OpenSearch/issues/1147

<img width="768" alt="Screenshot 2022-02-23 at 3 37 04 PM" src="https://user-images.githubusercontent.com/58062316/179953958-51749d8b-f59c-475d-8929-3a303332a29e.png">

PIT only takes data into account the moment a PIT is created. This implies on a lower level that none of the resources that are required to return the data from the initial request are modified or deleted. Segments are kept around, even though the segment might already have been merged away and is not needed for the live data set. This results in keeping more data around than just the live data set. More data means more segments, more file handles, and more heap to keep metadata from segments in the heap. 

## Scope
The document scope is only limited to Core OpenSearch changes needed to support PIT. *Design for OpenSearch Dashboards changes are pending.*

*2.1 Phase 1*

1. Supports create and delete PIT.
2. PIT expires after a keep-alive.
3. Use PIT Id in search request.
4. Supports delete all PITs
5. Monitoring PIT disk utilization.
6. Supports search slicing
7. OpenSearch Dashboard support. 

Phase 2 considerations <-- todo add link here

### Functional Requirements
1. Ability to submit a request to create a point in time for indices and get a PIT Id.
2. Ability to pass PIT Id in search request to query the segments kept alive as a part of the Point In Time. (Already implemented)
3. Ability to set a keep-alive to restrict the amount of time we retain the PIT shards’ segments (reader contexts)
4. Ability to set indices, preferences, routing, indices_options in a PIT in the same fashion as a _search.
5. Ability to submit a delete PIT.
6. Replace scroll API as the primary pagination solution (PIT + search_after, search slicing)

### Non functional Requirements
1. Idempotency —  Submitting the same search request with a given PIT Id should return the same result every time. If pagination is used, the n-th page should have the same result every time.
2. PIT should work unaffected by shard relocations.
3. Integration with OpenSearch Dashboards 
4. Integration with OS High Level Rest Client
5. Expired PITs should be eventually cleaned up.
6. User should be able to able to control PIT resource consumption through OpenSearch Settings.

### Point In Time Id

The identifier for a Point In Time, base64 URL encoded. It chiefly encapsulates the following information - for each shard that’s a part of the PIT it tells which node has the PIT segments i.e it has map of shard Id to node routing. Every node containing the copy of a given shard will not have segments marked for PIT.

```
SearchContextId:
    Map<ShardId, SearchContextIdForNode> shards;
```

```
SearchContextIdForNode:
    String node;
    ShardSearchContextId searchContextId; //unique id generated for each shard reader context.
```

### Point in time reader context

Point In Time Reader Context

For one copy of each shard in the PIT, a PIT Reader Context is created. A Reader Context holds a reference to a Lucene IndexSearcher.  An Index Searcher implements search over a single IndexReader.  IndexReader is a construct that  provides an interface for accessing a point-in-time view of an index. Any changes made to the index via IndexWriter will not be visible until a new IndexReader is opened. Hence by virtue of holding onto the searcher we are able to exclusively query the shards at that Point In Time.

The PIT Reader Context also holds information about the segments it retains for that shard. This helps us in monitoring the disk utilization of PITs.

```
PITReaderContext:
    ShardSearchContextId id
    List<Segments>
    TimeValue keepAlive
    TimeValue lastAccessedTime
    SearcherSupplier
    AbstractRefCounted refCounted;
    String pitId
```

#### What happens when we delete the PIT while a Point In Time Search is running:
When a PIT is marked for deletion, its reference count (refCounted in above POJO) is decremented by 1. But every search query that is using this PIT Reader Context will increment the reference by 1. The PIT Reader Context is only deleted when the refCounted hits 0. Hence even if a PIT is requested to be deleted, it is reaped from memory and resources held are released only once all searches are completed and no references are being held.

### Create Point In Time API

Unlike a scroll, by creating a dedicated PIT, we decouple the context from a single query and make it re-usable across arbitrary search requests by passing the PIT Id. We can achieve this by using the Create Point In Time API. 

```
POST "/<target indices>/point_in_time?*keep_alive*=1h&routing=&expand_wildcards=&preference=" 
```

```
{
    "id": "o463QQEPbXktaW5kZXgtMDAwMDAxFnNOWU43ckt3U3IyaFVpbGE1UWEtMncAFjFyeXBsRGJmVFM2RTB6eVg1aVVqQncAAAAAAAAAAAIWcDVrM3ZIX0pRNS1XejE5YXRPRFhzUQEWc05ZTjdyS3dTcjJoVWlsYTVRYS0ydwAA",
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "creation_time": 1658146050064
}
```

* Request
    * The path parameter is the target index/indices(all values eligible for /<target>/_search) are accepted.
    * The *keep_alive* parameter tells OpenSearch how long it should keep a point in time alive. It’s a *mandatory* query parameter. 
    * We limit the maximum keep-alive of a PIT using a cluster level setting "point_in_time.max_keep_alive" (defaults to 24h [keeping it the same as scrolls. Users can change it as required.])
    * We limit the number of open PIT contexts on each node using a node level setting "search.max_open_pit_context" (Defaults to 300. )
    * The optional parameters are **preference, routing, expand_wildcards** which have the same function as compared to passing them in a _search request.
    * The **allow partial results** parameter by default is false and is an optional parameter to allow PIT creation with partial failures.
* The result from the above request includes an id, which should be passed to the id of the pit parameter of a search request. Result will also contain _shards results as part of the response. Rest status will be based on the shards results.

#### Components
1. **RestCreatePITHandler** — The Rest endpoint for handling the create point in time request from the client. The rest handler is responsible for parsing the request, converting, delegating the request to the local transport handler, registering a response listener to be invoked once the response is ready to be invoked. 
2. **TransportCreatePITAction**
    1. The local transport handler for the request. To open readers for the shards which are a part of the PIT indices requested,  we make a morphed search request to assemble a list of Shard Targets (gives info about shard id+ index + node id) and only create and store contexts for those shards i.e. for only a single copy of each shard. Why are we calling this morphed? We don’t follow this query/fetch phase.
    2. In a normal search request, once we have a list of shards, we execute a phase on the shard - Shard Query Phase or Shard Fetch Phase.
    3.  **Phase 1 ( create PIT reader context )** -  Instead, in the case of PIT, we execute a single PIT-specific phase for each shard which calls SearchService.createPitReaderContext() on each node and sets a temporary keep-alive. ( 30 seconds ) 
    4. **Phase 2 ( save PIT ID in PIT reader context )** - Once we get the response of phase 1 from all the shards, we form the PIT ID and make another call to the same shards to store the PIT ID and also update the keep alive from the request. Creation time will also be broadcasted to all the shards along with PIT id and keep alive to keep it consistent.
    5. Will finalize which threadpool (generic or search) after running load tests.
3. **TransportSearchAction —** The plumbing logic to support the above mentioned single phase (create contexts Phase)search request has been done here. The PIT request’s indices, preferences, Once all the PIT shard contexts are opened, a search response is built from the shard phase results(list of per-shard PIT Context Id+node). The PIT Id is an encoding of these very results and is generated and returned as a part of this search response.

Why we need to store PIT ID  : 

* When a user creates a Point In Time, we generate a PIT Id which is a hash of information about which nodes the PIT shards lie on. This PIT id is only returned as a response to the create PIT API call.
* If the user couldn’t make note of the PIT Id or if he loses the PIT Id, he would have lost access to the Point In Time for the entirety of its duration and has no way to even delete the PIT without invoking a delete-all PIT’s command. 
* Users are not able to tag a set of PITs with labels to enable better classification(PITs can be created over different sets of indices at different times) and and management of point in times.

![createpit-1(1)](https://user-images.githubusercontent.com/58062316/179958426-add6785c-81e9-47cd-b8aa-7764807c6ed9.png)

#### Failure Scenarios:

**Phase 1**

* *Coordinator node fails*:
    *  request returns 5xx
    * Any PIT contexts created will get cleared asap.
* *All nodes or  all shards fail*:
    * request returns 5xx. no pit contexts created
* *Some nodes or some shards fail*:
    * request moves forward. 
    * failed shards/nodes wont have pit contexts.
    * PIT Id will be created from successful PIT context list

**Phase 2**

* If any of the shard update operation fails , we will fail the entire operation.
* *Coordinator node fails after broadcasting requests to some or all PIT contexts to update keep alive and PIT Id*
    *  request returns 5xx
    * updated PIT contexts will now become zombies and wont be reaped until keep alive elapses
* Coordinator node fails before broadcasting  update PIT Id, keep-alive request to any context:
    * request returns 5xx
    * PIT contexts will get cleared asap
* Coordinator node fails after updating all contexts and before returning response to client (I dont know if this is a plausible scenario)
    * request returns 5xx
    * updated PIT contexts will now become zombies and wont be reaped until keep alive elapses
* At least one PIT request fails to update PIT with PIT id, keep alive:
    * Coordinator node calls delete PIT.

###  Delete Point In Time API

PITs are automatically closed when the keep_alive is elapsed. However, keeping PITs has a cost; hence, they should be deleted as soon as they are no longer used in search requests. We may also delete a PIT and free the resources before its keep alive using the Delete Point In Time APIs.

```
`DELETE /_search/point_in_time

{
    "pit_id": [
        "o463QQEPbXktaW5kZXgtMDAwMDAxFkhGN09fMVlPUkVPLXh6MUExZ1hpaEEAFjBGbmVEZHdGU1EtaFhhUFc4ZkR5cWcAAAAAAAAAAAEWaXBPNVJtZEhTZDZXTWFFR05waXdWZwEWSEY3T18xWU9SRU8teHoxQTFnWGloQQAA",
        "o463QQEPbXktaW5kZXgtMDAwMDAxFkhGN09fMVlPUkVPLXh6MUExZ1hpaEEAFjBGbmVEZHdGU1EtaFhhUFc4ZkR5cWcAAAAAAAAAAAIWaXBPNVJtZEhTZDZXTWFFR05waXdWZwEWSEY3T18xWU9SRU8teHoxQTFnWGloQQAA"
    ]
}`

(or) Delete All :
`DELETE /_search/point_in_time/_all`
```

```
{
    "pits": [
        {
            "succeeded": true,
            "pitId": "o463QQEPbXktaW5kZXgtMDAwMDAxFkhGN09fMVlPUkVPLXh6MUExZ1hpaEEAFjBGbmVEZHdGU1EtaFhhUFc4ZkR5cWcAAAAAAAAAAAEWaXBPNVJtZEhTZDZXTWFFR05waXdWZwEWSEY3T18xWU9SRU8teHoxQTFnWGloQQAA"
        },
        {
            "succeeded": false,
            "pitId": "o463QQEPbXktaW5kZXgtMDAwMDAxFkhGN09fMVlPUkVPLXh6MUExZ1hpaEEAFjBGbmVEZHdGU1EtaFhhUFc4ZkR5cWcAAAAAAAAAAAIWaXBPNVJtZEhTZDZXTWFFR05waXdWZwEWSEY3T18xWU9SRU8teHoxQTFnWGloQQAA"
        }
    ]
}
```

**Request body:** (Required, string or array of strings) PIT IDs to be cleared. To clear all PIT IDs, use _all.
**Response body:** 
For each PIT Id requested to be deleted, we return a nested object with following fields

* **pitId :** pit id
    
* **succeeded :** whether the PIT is successfully deleted. Partial or complete failure are treated as failures.

<img width="1164" alt="delete PIT" src="https://user-images.githubusercontent.com/58062316/179959235-0e368222-2037-4289-887d-9a17fdc90057.png">

###  List all PIT API
This API returns all PIT IDs present in the ES cluster which can be used by the clients / UI 
```
GET "/_point_in_time/all
```

```
{
    "pits": [
        {
            "pitId": "o463QQEPbXktaW5kZXgtMDAwMDAxFnNOWU43ckt3U3IyaFVpbGE1UWEtMncAFjFyeXBsRGJmVFM2RTB6eVg1aVVqQncAAAAAAAAAAAEWcDVrM3ZIX0pRNS1XejE5YXRPRFhzUQEWc05ZTjdyS3dTcjJoVWlsYTVRYS0ydwAA",
            "creationTime": 1658146048666,
            "keepAlive": 6000000
        },
        {
            "pitId": "o463QQEPbXktaW5kZXgtMDAwMDAxFnNOWU43ckt3U3IyaFVpbGE1UWEtMncAFjFyeXBsRGJmVFM2RTB6eVg1aVVqQncAAAAAAAAAAAIWcDVrM3ZIX0pRNS1XejE5YXRPRFhzUQEWc05ZTjdyS3dTcjJoVWlsYTVRYS0ydwAA",
            "creationTime": 1658146050064,
            "keepAlive": 6000000
        }
    ]
}
```

**Response body:**
_nodes - Contains success and failure stats of all the nodes.
Id - PIT id
Creation time - PIT creation time ( same creation time is propagated to all reader contexts during create pit )
Keep Alive - Keep alive of PIT id

![listallpits(1)](https://user-images.githubusercontent.com/58062316/179959866-f6d661a3-1ff1-4a8d-bc6d-94de6e20d924.png)

**Failure scenarios**
If request to all nodes fail - request returns 5xx
If request to some nodes fail - success/failure info will be part of the response

### Point In Time Segments API

This API provides information about the disk utilization of a PIT. It returns low-level information about the Lucene segments a PIT is comprised of; similar to the cat segments API.

How do we get the information about the exact segments that are utilised by a PIT? When we open reader contexts for each of the PIT’s shards, we also store the segment info in the PIT reader Context. 

```
% curl "localhost:9200/_point_in_time/_segments/o4W9QAEFaW5kZXgWc09IcVlTY2pTNGVzNUIyYU12NklfUQAWaTIwRi1tRVBTTzJqZDBIN0JUUDdxdwAAAAAAAAAAARZPdFZNcVRFbFFENjNnMldDOW4wVHdRARZzT0hxWVNjalM0ZXM1QjJhTXY2SV9RAAA=?v"
index  shard prirep ip            segment generation docs.count docs.deleted  size size.memory committed searchable version compound
index1 0     r      10.212.36.190 _0               0          4            0 3.8kb        1364 false     true       8.8.2   true
index1 1     p      10.212.36.190 _0               0          3            0 3.7kb        1364 false     true       8.8.2   true
index1 2     r      10.212.74.139 _0               0          2            0 3.6kb        1364 false     true       8.8.2   true

```
**Request body: ** (Required, string) PIT ID

* **Components:**

1. **RestPITSegmentsHandler** — The Rest endpoint for handling the PIT segments request from the client. The rest handler is responsible for parsing the request, converting, delegating the request to the local transport handler, registering a response listener to be invoked once the response is ready to be invoked. From the PIT segments info received from the transport, builds a tabular response 
2. **TransportCreatePITAction —** The local transport handler for the request. It implements the abstraction, TransportBroadcastByNodeAction,  for transporting aggregated shard-level operations - in this case fetching segments info from the reader context -  in a single request per-node and executing the shard-level operations serially on the receiving node. Each shard-level operation can produce a result, these per-node shard-level results are aggregated into a single result to the coordinating node. These per-node results are aggregated into a single result (Result) to the client.

<img width="1190" alt="pit segments api" src="https://user-images.githubusercontent.com/58062316/179960367-ab52af74-90a0-4ef0-80b2-64eb9c7d9737.png">

**Comment by dreamer-89** [comment]
@bharath-techie : As today Sep 07, is the code freeze date for OpenSearch. Is there anything pending on this issue ?

**Comment by anasalkouz** [comment]
@bharath-techie is this still on track for 2.4 release? code freeze on 11/3
Is there anything pending? feel free to close it or change the label to 2.5

**Comment by dhruv16dhr** [comment]
@anasalkouz Yes this is on track for 2.4 release and will be closing it by next week (before 11/3)
@bharath-techie Please close this before 11/3
